php-checks/src/main/java/org/sonar/php/checks/WeakSSLProtocolCheck.java

```diff
@@ -106,39 +122,93 @@ public void visitFunctionCall(FunctionCallTree tree) {
   }
 
   private void checkStreamSSLConfig(ExpressionTree expressionTree) {
+    if (usages != null) {
+      var hasSensitiveReassignment = checkOptionsReassignment(usages, expressionTree);
+      if (hasSensitiveReassignment) {
+        // sensitive reassignment detected, no need to check the initial assignment
+        return;
+      }
+    }
+
     ExpressionTree config = getAssignedValue(expressionTree);
     if (!isArrayInitializer(config)) {
       return;
     }
     getProperty((ArrayInitializerTree) config, "SSL")
       .flatMap(sslConfig -> {
         if (isArrayInitializer(sslConfig)) {
-          return getProperty((ArrayInitializerTree) sslConfig, "crypto_method");
+          return getProperty((ArrayInitializerTree) sslConfig, CRYPTO_METHOD_KEY);
         }
         return Optional.empty();
       })
       .ifPresent(value -> checkStreamWeakProtocol(value, STREAM_CONTEXT_CREATE));
   }
 
+  private boolean checkOptionsReassignment(ReadWriteUsages usages, ExpressionTree expressionTree) {
+    var writes = usages.getWritesSorted(context().symbolTable().getSymbol(expressionTree));
+
+    var optionsSslAccesses = writes.stream()
+      .map(SyntaxToken::getParent)
+      .filter(tree -> tree instanceof VariableIdentifierTree variableIdentifierTree &&
+        variableIdentifierTree.getParent() instanceof ArrayAccessTree arrayAccessTree &&
+        isStringLiteralWithValue(arrayAccessTree.offset(), "ssl"))
+      .map(Tree::getParent)
+      .toList();
+
+    // In a sensitive case, `$options['ssl'] is either assigned with an array initializer or a value is directly written by an array key
+    // check if `$options['ssl']` is assigned with a sensitive array initializer
+    boolean hasOptionsSslSensitiveAssignment = optionsSslAccesses.stream()
+      .map(Tree::getParent)
+      .filter(Objects::nonNull)
+      .filter(tree -> tree.is(Tree.Kind.ASSIGNMENT))
+      .map(a -> ((AssignmentExpressionTree) a).value())
+      .filter(WeakSSLProtocolCheck::isArrayInitializer)
+      .findFirst()
+      .flatMap(c -> getProperty((ArrayInitializerTree) c, CRYPTO_METHOD_KEY))
+      .map(tree -> checkStreamWeakProtocol(tree, STREAM_CONTEXT_CREATE))
+      .orElse(false);
+
+    if (hasOptionsSslSensitiveAssignment) {
+      return true;
+    }
+
+    // check if `$options['ssl']['crypto_method']` is set directly
+    return optionsSslAccesses.stream()
+      .map(Tree::getParent)
+      .filter(a -> a instanceof ArrayAccessTree arrayAccessTree &&
+        isStringLiteralWithValue(arrayAccessTree.offset(), CRYPTO_METHOD_KEY))
+      .map(Tree::getParent)
+      .filter(Objects::nonNull)
+      .filter(tree -> tree.is(Tree.Kind.ASSIGNMENT))
+      .map(a -> ((AssignmentExpressionTree) a).value())
+      .findFirst()
+      .map(tree -> checkStreamWeakProtocol(tree, STREAM_CONTEXT_CREATE))
+      .orElse(false);
```

Comment:
Please move filters and mapping operations to small static methods. This will make it easier to read such heavy stream operations.